overlord/devicestate/devicemgr.go

```diff
@@ -1877,74 +1877,92 @@ func (m *DeviceManager) Systems() ([]*System, error) {
-	}
-	kernelInfo, err := snap.ReadInfoFromSnapFile(snapf, kernelSnap.SideInfo)
+	// System information
+	sys, snapInfos, seedSnaps, err := m.loadSystemAndEssentialSnaps(wantedSystemLabel, []snap.Type{snap.TypeKernel, snap.TypeGadget})
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	if kernelInfo.SnapType != snap.TypeKernel {
-		return nil, nil, nil, fmt.Errorf("cannot use kernel info, expected kernel but got: %v", kernelInfo.SnapType)
-	}
 
-	// gadget info needed to check encryptionSupport and gadget contraints
-	gadgetSnap := s.EssentialSnaps()[1]
-	if gadgetSnap.Path == "" {
-		return nil, nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
-	}
-	snapf, err = snapfile.Open(gadgetSnap.Path)
+	// Gadget information
+	snapf, err := snapfile.Open(seedSnaps[snap.TypeGadget].Path)
 	if err != nil {
 		return nil, nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
 	}
-	gadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
-	}
-	gadgetInfo, err := gadget.InfoFromGadgetYaml(gadgetYaml, sys.Model)
+	gadgetInfo, err := gadget.ReadInfoFromSnapFileNoValidate(snapf, sys.Model)
 	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("reading gadget information: %v", err)
 	}
 
-	encInfo, err := m.encryptionSupportInfo(sys.Model, kernelInfo, gadgetInfo)
+	// Encryption details
+	encInfo, err := m.encryptionSupportInfo(sys.Model, snapInfos[snap.TypeKernel], gadgetInfo)
 	if err != nil {
 		return nil, nil, nil, err
 	}
+
+	// Make sure gadget is valid for model and available encryption
 	opts := &gadget.ValidationConstraints{
 		EncryptedData: encInfo.StorageSafety == asserts.StorageSafetyEncrypted,
 	}
 	if err := gadget.Validate(gadgetInfo, sys.Model, opts); err != nil {
 		return nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
 	}
 
-	return sys, gadgetInfo, &encInfo, nil
+	return sys, gadgetInfo, &encInfo, err
+}
+
+// loadSystemAndEssentialSnaps loads information for the given label, which
+// includes system, gadget information, gadget and kernel snaps info,
+// and gadget and kernel seed snap info.
+func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (
+	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
+
+	// get current system as input for loadSeedAndSystem()
+	systemMode := m.SystemMode(SysAny)
+	currentSys, _ := currentSystemForMode(m.state, systemMode)
+
+	s, sys, err := loadSeedAndSystem(wantedSystemLabel, currentSys)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// 2. get the gadget volumes for the given system-label
+	perf := &timings.Timings{}
+	if err := s.LoadEssentialMeta(types, perf); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot load essential snaps metadata: %v", err)
+	}
+	// EssentialSnaps is always ordered, see asserts.Model.EssentialSnaps:
```

Comment:
(nitpick) We don't need this comment anymore, do we? We do not rely on ordering as we use the "seedSnap.EssentialType" now?